### PR TITLE
`step(::AbstractUnitRange{Bool})` should return `Bool`

### PR DESCRIPTION
**Pull request by putianyi889**
The issue was introduced by #27302 , as
```julia
julia> true-false
1
```

By definitions below, `AbstractUnitRange{Bool} <: OrdinalRange{Bool, Bool}` whose step type is `Bool`.

https://github.com/JuliaLang/julia/blob/da74ef1933b12410b217748e0f7fbcbe52e10d29/base/range.jl#L280-L299

### CODE REVIEW

File: base/range.jl [hunk 1]
```diff
@@ -711,6 +711,7 @@ julia> step(range(2.5, stop=10.9, length=85))
 """
 step(r::StepRange) = r.step
 step(r::AbstractUnitRange{T}) where {T} = oneunit(T) - zero(T)
+step(r::AbstractUnitRange{T}) where {T} = true
```
Comment (jishnub) [inline base/range.jl:714]: Isn't this overwriting the method on line 713?

File: base/range.jl [hunk 2]
```diff
@@ -711,7 +711,7 @@ julia> step(range(2.5, stop=10.9, length=85))
 """
 step(r::StepRange) = r.step
 step(r::AbstractUnitRange{T}) where {T} = oneunit(T) - zero(T)
```
Comment (jishnub) [inline base/range.jl:713]: I wonder why this definition is used here? The docstring of `AbstractUnitRange` states that the step is `oneunit(T)`, which seems to differ from this definition.

Comment (putianyi889) [reply base/range.jl:713]: ask https://github.com/JuliaLang/julia/pull/27302

**Review by jishnub — APPROVED** [review]

**Comment by mbauman** [comment]
Could you describe why this was an issue for you?  It's hard for me to understand why this definition was "wrong" for `Bool` in particular.

If this is the wrong approach, perhaps we should reconsider the general behavior, too.  Or if this was an issue with the docs of `AbstractUnitRange`, then perhaps that should be clarified.

**Comment by mbauman** [comment]
In other words, #27302 works _in spite of_ breaking this promise to the supertype — not just for Bool, but for specific types it added and tested, too.

```
julia> steptype(::OrdinalRange{<:Any,S}) where S = S
steptype (generic function with 1 method)

julia> steptype(Position(1):Position(10))
Position

julia> step(Position(1):Position(10))
Displacement(1)
```

So unless this manifested as a real and pressing user-level bug for which it'd make sense to do a quick-n-dirty patch, I think we should consider how we address this more holistically.

**Comment by putianyi889** [comment]
> not just for Bool, but for specific types it added and tested, too.

IMO it's the types it added that broke conventions. Compatibility with custom types should be addressed without breaking `Base`, after all, the interface can always be customized for special types. Bottomline is that the documentation and the implementation in `Base` should be consistent.

**Comment by mbauman** [comment]
My core question is exactly that — is this PR addressing a docs problem or is it a functionality problem?

I'd even question if we need to have two parameters on the `OrdinalRange` abstract type. We don't touch that second parameter in **any** of the builtin methods, and [even in the ecosystem](https://juliahub.com/ui/Search?type=code&q=::OrdinalRange{) everyone seems to ignore it.  Of course, we can't remove it now — that'd be _very_ breaking.  But perhaps it should be documented to be vestigial?

Maybe this PR is still a good thing to do; I'm just not sure if it _functionally_ changes things (let alone for better or worse).

Any which way, I think the best way to tackle the inconsistency is by cleaning up the docs to reflect reality.  For example, they're already talking about `Date` "naturally" working, but it doesn't because there's not a `oneunit(Dates.today())`.

**Comment by putianyi889** [comment]
> My core question is exactly that — is this PR addressing a docs problem or is it a functionality problem?

Could be docs problem, but based on the type declaration, could also be functionality problem.

> I'd even question if we need to have two parameters on the OrdinalRange abstract type.

Maybe yes. It'll be nice for custom types, and even `StepRange` can have different types for start and step.

> I think the best way to tackle the inconsistency is by cleaning up the docs to reflect reality.

I suppose it's the core devs' duty? The reality is quite complicated and may not fully follow the core team's principles. A summary needs a deep dig into the codes, and more issues could be found.

**Comment by mbauman** [comment]
My point is that this PR changes some functionality. I'm trying to understand in what circumstances — if any — this change is noticeable.

I'm also trying to understand what led to the problem here in the first place — and if there might be any better alternatives.

To be clear: it's very helpful that you flagged this, and thanks for the PR @putianyi889! 

> > I'd even question if we need to have two parameters on the OrdinalRange abstract type.
> 
> Maybe yes. It'll be nice for custom types, and even `StepRange` can have different types for start and step.

Removing/deprecating/ignoring the second parameter on `OrdinalRange` wouldn't affect `StepRange` or any other subtypes at all — in fact, it gives them _more latitude_ as they needn't worry about knowing what the step type will be at definition time.

In fact, I'm not sure _what_ you'd do with the type of the step in the first place — and whether you'd expect this to be the internal step type or the `step` return type.  In fact, floating point ranges lie about the true type (and value!) of their `step`.  But that's all https://github.com/JuliaLang/julia/issues/40466.

**Comment by putianyi889** [comment]
> I'm also trying to understand what led to the problem here in the first place — and if there might be any better alternatives.
I think it's the lack of definitive standards of those abstract types. For example, there is no definition on what `OrdinalRange` is exactly and what its interfaces do. Without them, people can have different interpretations.

> I'm trying to understand in what circumstances — if any — this change is noticeable.

Currently, the issue causes unexpected eltype lifting:
```julia
julia> (false:true:true) + (Int8(0):Int8(1)) |> eltype
Int8

julia> (false:true) + (Int8(0):Int8(1)) |> eltype
Int64
```
It's because the step type of `false:true` becomes `Int64`. It's not a problem in production (that I know of), but it's what I encountered when writing tests.

**Comment by mbauman** [comment]
Oh nice, that's exactly the kind of higher-level behavior I was curious about.  Yes, then I think it makes sense to "patch up" these `Bool` ranges like this.  Could you add that addition |> eltype example as another test?

File: test/ranges.jl [hunk 3]
```diff
@@ -65,6 +65,8 @@ using .Main.OffsetArrays
 
     unitrangeerrstr = "promotion of types Char and Char failed to change any arguments"
     @test_throws unitrangeerrstr UnitRange('a', 'b')
+
+    @test step(false:true) === true # PR 56405
```
Comment (mbauman) [inline test/ranges.jl:69]:
```suggestion
    @test step(false:true) === true # PR 56405
    @test eltype((false:true) + (Int8(0):Int8(1))) === Int8
    @test eltype((false:true:true) + (Int8(0):Int8(1))) === Int8
```

**Review by mbauman — APPROVED** [review]

**Pull request merged**